File: Problems/Algorithms/322. Coin Change/CoinChange.kt
class Solution {
    fun coinChange(coins: IntArray, amount: Int): Int {
        val dp = IntArray(amount+1) { -1 }
        dp[0] = 0
        
        coins.sort()
        for (i in 1..amount) {
            for (coin in coins) {
                if (i == coin) {
                    dp[i] = 1
                    break
                } else if (i > coin) {
                    val temp = if (dp[i-coin] == -1) -1 else dp[i-coin] + 1
                    dp[i] = if (dp[i] == -1) temp else if (temp != -1) minOf(dp[i], temp) else dp[i]
                }
            }
        }
        
        return dp[amount]
    }
}

File: Problems/Algorithms/322. Coin Change/coin_change.py
class Solution:
    def coinChange(self, coins: List[int], amount: int) -> int:
        dp = [-1] * (amount+1)
        dp[0] = 0

        for i in range(1, amount+1):
            curr = [dp[i-j] for j in coins if j <= i and dp[i-j] != -1]
            if curr:
                dp[i] = min(curr) + 1

        return dp[amount]

